### DISCUSSION TITLE
Code too big to fit??

### DISCUSSION THREAD

**Discussion by mechatroniks-git** [SpenceKonde/megaTinyCore#574] [Q&A] [answered]
I'm trying to decipher this. Does it mean my code is being truncated to fit in the flash memory?

```
C:\Users\win10\AppData\Local\Temp\ccwgySEj.ltrans0.ltrans.o: In function `UartClass::_poll_tx_data_empty()':
<artificial>:(.text+0x2f2): relocation truncated to fit: R_AVR_13_PCREL against `no symbol'
C:\Users\win10\AppData\Local\Temp\ccwgySEj.ltrans0.ltrans.o: In function `u8x8_byte_arduino_hw_i2c':
<artificial>:(.text+0x1cd2): relocation truncated to fit: R_AVR_13_PCREL against `no symbol'
collect2.exe: error: ld returned 1 exit status
exit status 1
Error compiling for board ATtiny3224/1624/1614/1604/824/814/804/424/414/404/214/204.
```

**Comment by mechatroniks-git** [comment]
Arduino 1.8.19, megatinycore 2.5.2
Here is my sketch running happy

```
Sketch uses 10914 bytes (66%) of program storage space. Maximum is 16384 bytes.
Global variables use 858 bytes (41%) of dynamic memory, leaving 1190 bytes for local variables. Maximum is 2048 bytes.
```

Once I add the following line to setup(), I get the above error. 
'''Serial.begin(115200);'''

printf() as minimal, default or full - no difference.

**Comment by SpenceKonde** [comment]
Fuck!

No, it means I screwed up and my testing wasn't sufficient, and I have to change and reissue emergency fixes for DxCore and megaTinyCore.

**Comment by SpenceKonde** [comment]
This is now available in the board manager release

**Reply by mechatroniks-git** [reply]
Thanks!

**Reply by swevedg** [reply]
I have a similar problem - Arduino 1.8.19, megatinycore 2.5.3
Program for ATtiny1624 builds/works fine with Serial.begin (115200) and four Serial.print statements - adding one more Serial.print fails. Biggest build which works uses 12285 bytes storage and 381 bytes dynamic memory.
Error message is identical with that posted by OP.

**Comment by mechatroniks-git** [comment]
@swevedg I'm still seeing it also, noted @ #575

**Reply by swevedg** [reply]
On the off-chance that it might work, I tried Arduino IDE 2.0.0-rc3, megatinycore 2.5.3.
It successfully compiles with all my Serial.print statements included (There are lots of them). Code size is now 13073 bytes.
I am reluctant to change compilers before the new one is officially released but I think I will now try it on the hardware and see if that works.

**Reply by mechatroniks-git** [reply]
@swevedg I had IDE 2.0.0-rc2 installed and just tried it, still getting same errors. I'll try updating IDE.

**Reply by swevedg** [reply]
I had no luck programming the hardware - the IDE 2.0.0-rc3 does not recognise the programmer - "Serial UPDI with 4.7k resistor..."
I've run out of time now, I'll try again tomorrow.

**Reply by SpenceKonde** [reply]
The compiler toolchain used should be the one I supply either way, so that's weird.... There is definitely a bug in 2.5.3 which may or may not be the cause of this, but I don;t have any code that reproduced the issue.

**Reply by swevedg** [reply]
I discovered this problem the day before yesterday. I found many references to the error message and eventually found two references that explained the cause and a fix and which seemed reasonable to me:

Cause: https://www.stix.id.au/wiki/AVR_relocation_truncations_workaround
Fix: https://forum.arduino.cc/t/how-to-add-compile-options-to-arduino-ide-1-8-13/692640/2

I was about to try and figure out how to apply the fix when I saw this thread. I am not to confident as to my ability to fiddle with the inner workings of the IDE but I may eventually have to try.

**Comment by SpenceKonde** [comment]
It should be sorted out by 2.5.4?


____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by swevedg** [reply]
Yes, it looks like it is sorted out. My code now builds with all of the serial port calls included again. I now have another, probably unrelated, problem which I need to investigate. 
As a matter of interest, was the problem in any way related to the RCALL/RJMP problem?

**Reply by swevedg** [reply]
My apologies - I meant also to say thank you very much for the responses.

**Comment by SpenceKonde** [comment, marked as answer]
Yeah, I was testing flash size with ifdefs in a block of inline assembly,
because I figured out that I needed to use jmp not rjmp on things with more
than 8k of flash but had 0x in front of the size in the test x_x

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***
